net/instaweb/rewriter/insert_ga_filter.cc

```diff
@@ -141,56 +199,82 @@ void InsertGAFilter::StartElementImpl(HtmlElement* element) {
-        buffer_.find(".google-analytics.com/urchin.js") != GoogleString::npos));
-}
-
-// Running an experiment: add in the information as the slot 1 custom variable.
-// TODO(nforman): Change this to be a label on track_timings
-// data when the track_timings api goes live (maybe).
-GoogleString InsertGAFilter::ConstructExperimentSnippet() const {
-  GoogleString experiment = "";
-  if (driver()->options()->running_experiment()) {
-    int experiment_state = driver()->options()->experiment_id();
-    if (experiment_state != experiment::kExperimentNotSet &&
-        experiment_state != experiment::kNoExperiment) {
-      experiment = StringPrintf(kExperimentSnippetFmt,
-          driver()->options()->experiment_ga_slot(),
-          driver()->options()->ToExperimentString().c_str());
+InsertGAFilter::AnalyticsStatus InsertGAFilter::FindSnippetInScript(
+    const GoogleString& s) {
+  if (!seen_ga_js_ &&
+      s.find("google-analytics.com/ga.js") != GoogleString::npos) {
+    seen_ga_js_ = true;
+  }
+  if (s.find(ga_id_) == GoogleString::npos) {
+    return kNoSnippetFound;
+  }
+  if (s.find(".google-analytics.com/urchin.js") != GoogleString::npos) {
+    return kUnusableSnippetFound;  // urchin.js is too old.
+  } else if (s.find(".google-analytics.com/ga.js") != GoogleString::npos) {
+    if (s.find("_setAccount") != GoogleString::npos) {
+      return kGaJs;  // Asynchronous ga.js
     }
+    return kUnusableSnippetFound;
+  } else if (seen_ga_js_ &&
+             s.find("_getTracker") != GoogleString::npos &&
+             s.find("_trackPageview") != GoogleString::npos) {
+    // Synchronous ga.js was split over two script tags: first one to do the
+    // loading then one to do the initiailization and page tracking.  We want to
+    // process the second one.
+    return kGaJs;  // Syncronous ga.js
+  } else if (s.find(".google-analytics.com/analytics.js")) {
+    if (RE2::PartialMatch(s, kSendPageviewRegexp)) {
+      return kAnalyticsJs;
+    }
+    return kUnusableSnippetFound;
   }
-  return experiment;
+  return kNoSnippetFound;
 }
 
 void InsertGAFilter::AddScriptNode(HtmlElement* current_element,
-                                   GoogleString text,
-                                   bool insert_immediately_after_current) {
-  HtmlElement* script_element = driver()->NewElement(current_element,
-                                                     HtmlName::kScript);
+                                   const GoogleString& text,
+                                   const GoogleString& url,
+                                   bool append_child) {
+  DCHECK(text.empty() != url.empty()) <<
+      "Exactly one of text/url should be set.";
+
+  HtmlElement* parent;
+  if (append_child) {
+    parent = current_element;
+  } else {
+    parent = current_element->parent();
+    if (parent == NULL) {
+      LOG(INFO) << "Null parent in insert_ga can't insert following node.";
```

Comment:
This could happen outside of the operator's control if pagespeed was running in a proxy or something, which makes me wary about calling it a warning.